yeps/yep-312.md

```diff
@@ -0,0 +1,169 @@
+The functional form is daemon specific; parameters for the functional form should be specified in state or config, whichever the developer sees as appropriate.
+The inverse of the function, `to_native`, is also specified.
+
+The trait introduces one explicit functional parameter, the state variable `reference_position`, which establishes the null position (0 in the Schemes above) in the native coordinate system.
+
+Transform parameters are subject to change, so native positions are more robust.  As such, state is recorded through native units.
+In contrast, client calls to the `position` property will report transformed coordinates.
+
+# Specification
+
+This trait requires `has-limits` (<a href="../303">YEP-303</a>).
+
+## State
+
+### native_reference_position (float)
+
+The reference position, expressed in native coordinates.
+The transformed function treats the reference position as the "zero" point.
+Default is `0.0` (i.e. reference position is the same as hardware null).
+
+## Messages
+
+### to_transformed
+
+request : float
+
+response : float
+
+Convert a native coordinates to transformed.
+
+### to_native
+
+request : float
+
+response : float
+
+Convert a transformed coordinates to native.
+The inverse of `to_transformed`.
+
+### set_native_reference
+
+request : float
+
+Set the reference position in native coordinates.
+Reference position need not be within hardware limits.
+
+### get_native_reference
+
+response : float
+
+Get the reference position in native coordinates.
+
+### get_native_position
+
+response : float
+
+Get the current position in native coordinates.
+
+### set_native_position
+
+request : float
+
+Set the current position in native coordinates.
+
+### get_native_limits
+
+response : {"type": "array", "values" : ["float]}
+
+Returns limits in the relative coordinate system.
+
+### get_native_units
+
+response : "string"
+
+Get the units of native coordinates.
+
+## Properties
+
+### native_reference_position
+
+* getter : "get_reference_position"
+* setter : "set_reference_position"
+* units_getter : "get_native_units"
+* control_kind : "hinted"
+* record_kind : "metadata"
+
+### native_position
+
+* getter : "get_native_position"
+* setter : "set_native_position"
+* limits_getter : "get_native_limits"
+* units_getter : "get_native_units"
+* control_kind : "hinted"
+* record_kind : "metadata"
+
+### native_destination
+
+* getter : "get_native_destination"
+* units_getter : "get_native_units"
+* control_kind : "hinted"
+* record_kind : "metadata"
+
+# Backwards Compatibility
+
+This trait has no effect on `has-limits` when `reference_position = 0.0` and `to_transformed(x) == to_native(x) == x`.
+
+In state, all values relating to position (`position`, `destination`, `native_reference_position` and `hw_limits`) are stored in native coordinates.
+This is so that state values will remain valid even when transforms change.
+
+# Rejected Ideas
+
```

Comment:
some mention of the ideas we went through to get here may be useful context here